File: core/ts/test/keytext.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import { encodeKey, decodeKey } from "../src/index.js";

const u8 = (n: number, len = 32): Uint8Array => new Uint8Array(len).fill(n);

test("encode fixed form", () => {
  assert.equal(encodeKey(u8(0x11)), "ed25519:" + "11".repeat(32));
});

test("round-trip", () => {
  const key = Uint8Array.from({ length: 32 }, (_, i) => i);
  assert.deepEqual(decodeKey(encodeKey(key)), key);
});

test("decode accepts uppercase hex", () => {
  assert.deepEqual(decodeKey("ed25519:" + "AB".repeat(32)), decodeKey("ed25519:" + "ab".repeat(32)));
});

test("decode rejects missing prefix", () => {
  assert.throws(() => decodeKey("11".repeat(32)));
});

test("decode rejects odd-length body", () => {
  assert.throws(() => decodeKey("ed25519:111"));
});

test("decode rejects non-hex body", () => {
  assert.throws(() => decodeKey("ed25519:" + "zz".repeat(32)));
});

test("decode rejects wrong decoded length", () => {
  assert.throws(() => decodeKey("ed25519:" + "11".repeat(31)));
  assert.throws(() => decodeKey("ed25519:" + "11".repeat(33)));
});

File: core/go/keytext/keytext_test.go
package keytext

import (
	"bytes"
	"strings"
	"testing"
)

func TestEncodeFixedForm(t *testing.T) {
	got := EncodeKey(bytes.Repeat([]byte{0x11}, 32))
	want := "ed25519:" + strings.Repeat("11", 32)
	if got != want {
		t.Fatalf("encode mismatch:\n got %q\nwant %q", got, want)
	}
}

func TestRoundTrip(t *testing.T) {
	key := make([]byte, 32)
	for i := range key {
		key[i] = byte(i)
	}
	text := EncodeKey(key)
	back, err := DecodeKey(text)
	if err != nil {
		t.Fatalf("decode failed: %v", err)
	}
	if !bytes.Equal(back, key) {
		t.Fatalf("round-trip mismatch:\n got %x\nwant %x", back, key)
	}
}

func TestDecodeAcceptsUppercaseHex(t *testing.T) {
	lower, errL := DecodeKey("ed25519:" + strings.Repeat("ab", 32))
	upper, errU := DecodeKey("ed25519:" + strings.Repeat("AB", 32))
	if errL != nil || errU != nil {
		t.Fatalf("decode errors: lower=%v upper=%v", errL, errU)
	}
	if !bytes.Equal(lower, upper) {
		t.Fatalf("uppercase/lowercase mismatch")
	}
}

func TestDecodeRejectsBadInput(t *testing.T) {
	cases := map[string]string{
		"missing prefix": strings.Repeat("11", 32),
		"odd length":     "ed25519:111",
		"non-hex":        "ed25519:" + strings.Repeat("zz", 32),
		"too short":      "ed25519:" + strings.Repeat("11", 31),
		"too long":       "ed25519:" + strings.Repeat("11", 33),
	}
	for name, in := range cases {
		if _, err := DecodeKey(in); err == nil {
			t.Errorf("%s: expected an error for %q", name, in)
		}
	}
}
